### DISCUSSION TITLE
RTR_ExecuteActiveResponderCommand 400 Errors

### DISCUSSION THREAD

**Discussion by brianp9906** [CrowdStrike/falconpy#466] [Q&A] [answered]
I've been pulling my hair out because no matter how I pass the commands to Falcon API, it always comes back with a 400. 
Any idea why this command fails?

`                          BODY = {
                                "base_command": "runscript",
                                "command_string": "runscript -Raw=```ls /```",
                                "session_id": resource['session_id']
                            }
                            response = falcon_RTR_Session.RTR_ExecuteActiveResponderCommand(body=BODY)
                            print(json.dumps(response, indent=4)) 
`

result is:
`{
    "status_code": 400,
    "headers": {
        "Content-Encoding": "gzip",
        "Content-Length": "214",
        "Content-Type": "application/json",
        "Date": "Wed, 01 Dec 2021 18:07:53 GMT",
        "Strict-Transport-Security": "max-age=31536000; includeSubDomains",
        "X-Cs-Region": "us-1",
        "X-Cs-Traceid": "2d943f10-31cb-4cff-b083-f23d504a9a7c",
        "X-Ratelimit-Limit": "6000",
        "X-Ratelimit-Remaining": "5993"
    },
    "body": {
        "meta": {
            "query_time": 0.00539614,
            "powered_by": "empower-api",
            "trace_id": "2d943f10-31cb-4cff-b083-f23d504a9a7c"
        },
        "resources": [],
        "errors": [
            {
                "code": 40014,
                "message": "Unrecognized flag found: Raw, value: ls /"
            }
        ]
    }
}
`

I tried using a script name and that fails with a 400 status code too.  Python wants to add double backslashes so that doesnt work either. 

`'{"base_command": "runscript", "command_string": "runscript -CloudFile=\\"MyScriptNameToRun\\" -CommandLine=\\"\\""}'
`

Result is 'Command is not valid'

**Comment by jshcodes** [comment]
Hi @brianp9906!

I'm checking out the payload you provided now.  Question - Are you using the Uber Class or the RealTimeResponse Service Class?

**Comment by jshcodes** [comment, marked as answer]
Here is an example pulling a directory listing using the __RealTimeResponse Service Class__ and [Body Payload Abstraction](https://falconpy.io/Usage/Payload-Handling.html#body-payload-abstraction) and an example doing the same thing passing a raw body payload like you demonstrate.

```python
import os
from falconpy import RealTimeResponse

# This example leverages Body Payload Abstraction to perform the directory listing.
falcon = RealTimeResponse(client_id=os.environ["FALCON_CLIENT_ID"],
                          client_secret=os.environ["FALCON_CLIENT_SECRET"]
                          )

DEVICE_AID = "DEVICE_AID_HERE"

# I'm using Operation ID syntax since that's what your question used.
session_id = falcon.RTR_InitSession(device_id=DEVICE_AID)["body"]["resources"][0]["session_id"]

response = falcon.RTR_ExecuteActiveResponderCommand(base_command="runscript",
                                                    command_string="ls /",
                                                    session_id=session_id
                                                    )

cloud_request_id = response["body"]["resources"][0]["cloud_request_id"]
completed = False
while not completed:
    check_result = falcon.RTR_CheckActiveResponderCommandStatus(cloud_request_id=cloud_request_id,
                                                                sequence_id=0
                                                                )

    completed = check_result["body"]["resources"][0]["complete"]

# Display the result
print(check_result["body"]["resources"][0]["stdout"])


# This example does the same thing, but uses a raw body payload
# We will re-use our existing session to the target host
body_payload = {
    "session_id": session_id,
    "base_command": "runscript",
    "command_string": "ls /"
}

response = falcon.RTR_ExecuteActiveResponderCommand(body=body_payload)

cloud_request_id = response["body"]["resources"][0]["cloud_request_id"]
completed = False
while not completed:
    check_result = falcon.RTR_CheckActiveResponderCommandStatus(cloud_request_id=cloud_request_id,
                                                                sequence_id=0
                                                                )
    completed = check_result["body"]["resources"][0]["complete"]

# Display the result
print(check_result["body"]["resources"][0]["stdout"])
```

**Reply by brianp9906** [reply]
Thanks @jshcodes.  This was super helpful.  The challenging part is that the examples from https://falconpy.io/Service-Collections/Real-Time-Response.html werent clear for syntax on the parameters of using runscript so that helped a bunch. 

I was able to run my custom script like this:
```python
BODY = {
    "base_command": "runscript",
    "command_string": "runscript -CloudFile=\"MyScriptName\" -CommandLine=\"\"",
    "session_id": resource['session_id']
}
command_response = falcon_RTR_Session.RTR_ExecuteActiveResponderCommand(body=BODY)
print(json.dumps(command_response, indent=4)) 
```

**Reply by jshcodes** [reply]
Excellent!

**Reply by jshcodes** [reply]
We also have some more RTR examples [here](https://github.com/CrowdStrike/falconpy/tree/main/samples/rtr).
